src/Schema/DefinitionBuilder.php

```diff
@@ -468,31 +536,90 @@ protected function getDeprecationReason(NodeInterface $node): ?string
+    }
+
+    /**
+     * @param string $typeName
+     * @return NamedTypeInterface
+     * @throws InvalidArgumentException
+     */
+    protected function getTypeFromCache(string $typeName): NamedTypeInterface
+    {
+        return $this->getFromCache($this->createTypeCacheKey($typeName));
+    }
+
+    /**
+     * @param string             $typeName
+     * @param NamedTypeInterface $type
+     * @return bool
+     * @throws InvalidArgumentException
+     */
+    protected function setTypeInCache(string $typeName, NamedTypeInterface $type): bool
+    {
+        return $this->setInCache($this->createTypeCacheKey($typeName), $type);
+    }
+
+    /**
+     * @param string $typeName
      * @return string
      */
-    protected function getCachePrefix(): string
+    protected function createTypeCacheKey(string $typeName): string
     {
-        return self::CACHE_PREFIX;
+        return $typeName . self::TYPE_CACHE_SUFFIX;
     }
-}
 
-/**
- * @param TypeInterface                        $innerType
- * @param NamedTypeInterface|TypeNodeInterface $inputTypeNode
- * @return TypeInterface
- * @throws InvariantException
- * @throws InvalidTypeException
- */
-function buildWrappedType(TypeInterface $innerType, TypeNodeInterface $inputTypeNode): TypeInterface
-{
-    if ($inputTypeNode instanceof ListTypeNode) {
-        return newList(buildWrappedType($innerType, $inputTypeNode->getType()));
+    /**
+     * @param string $directiveName
+     * @return bool
+     * @throws InvalidArgumentException
+     */
+    protected function isDirectiveInCache(string $directiveName): bool
+    {
+        return $this->isInCache($this->createDirectiveCacheKey($directiveName));
     }
 
-    if ($inputTypeNode instanceof NonNullTypeNode) {
-        $wrappedType = buildWrappedType($innerType, $inputTypeNode->getType());
-        return newNonNull(assertNullableType($wrappedType));
+    /**
+     * @param string $directiveName
+     * @return Directive
+     * @throws InvalidArgumentException
+     */
+    protected function getDirectiveFromCache(string $directiveName): Directive
+    {
+        return $this->getFromCache($this->createDirectiveCacheKey($directiveName));
     }
 
-    return $innerType;
+    /**
+     * @param string    $directiveName
+     * @param Directive $directive
+     * @return bool
+     * @throws InvalidArgumentException
+     */
+    protected function setDirectiveInCache(string $directiveName, Directive $directive): bool
+    {
+        return $this->setInCache($this->createDirectiveCacheKey($directiveName), $directive);
+    }
+
+    /**
+     * @param string $directiveName
+     * @return string
+     */
+    protected function createDirectiveCacheKey(string $directiveName): string
+    {
+        return $directiveName . self::DIRECTIVE_CACHE_SUFFIX;
+    }
+
+    /**
+     * @return string
+     */
+    protected function getCachePrefix(): string
+    {
+        return self::CACHE_PREFIX;
+    }
```

Comment:
I'd consider adding a class for handling the interaction with the cache, e.g. `SchemaCache`. What do you think @fubhy?